src/os/bluestore/bluestore_dump.cc

```diff
@@ -0,0 +1,438 @@
+
+  //TODO - improve on error with decoding size
+  ceph_assert(p.get_offset() == v_size);
+  return true;
+}
+
+
+bool Dump::dump_object(const std::string& key, const bufferlist& v)
+{
+  ghobject_t oid;
+  int r = BlueStore::objkey_to_oid(key, &oid);
+  if (r != 0) {
+    cout << "failed to decode '" + url_escape(key) + "' error code=" + std::to_string(r) << std::endl;
+    return false;
+  }
+  f->open_object_section("object");
+  f->open_object_section("id");
+  f->dump_string("hobj", oid.hobj.to_str());
+  f->dump_int("shard", oid.shard_id);
+  if (oid.generation != ghobject_t::NO_GEN)
+    f->dump_unsigned("gen", oid.generation);
+  else
+    f->dump_int("gen", -1);
+  f->close_section();
+  dump_object(v);
+  f->close_section();
+  return true;
+}
+
+
+bool Dump::dump_extent(const std::string& key, const bufferlist& v)
+{
+  ghobject_t oid;
+  uint32_t offset;
+  int r = BlueStore::extentkey_to_oid(key, &oid, &offset);
+  if (r != 0) {
+    cerr << "failed to decode '" + url_escape(key) + "' error code=" + std::to_string(r) << std::endl;
+    return false;
+  }
+  f->open_object_section("extent");
+  f->open_object_section("id");
+  f->dump_string("hobj", oid.hobj.to_str());
+  f->dump_int("shard", oid.shard_id);
+  if (oid.generation != ghobject_t::NO_GEN)
+    f->dump_unsigned("gen", oid.generation);
+  else
+    f->dump_int("gen", -1);
+  f->close_section();
+  f->dump_unsigned("offset", offset);
+
+  decode_some(v);
+  f->close_section();
+  return true;
+}
+
+
+int Dump::list_objects(const std::string& format,
+		       const std::string& selector,
+		       const std::string& trimmer) {
+  KeyValueDB::Iterator obj_it = db->get_iterator("O");
+  obj_it->seek_to_first();
+  shared_ptr<Formatter> actual_formatter(Formatter::create(format));
+  if (actual_formatter.get() == nullptr) {
+    std::cout << "Invalid format '" << format << "'" << std::endl;
+    return -1;
+  }
+  shared_ptr<Filter> ff;
+  shared_ptr<Buffer> bf;
+  shared_ptr<Trimmer> tf;
+  Formatter* buffer;
+  bool matched;
+
+  if (!selector.empty()) {
+    std::string error;
+    ff.reset(Filter::create());
+    bool b = ff->setup(selector, error);
+    if (!b) {
+      std::cout << error << std::endl;
+      return -1;
+    }
+    bf.reset(Buffer::create());
+  }
+
+  if (!trimmer.empty()) {
+    std::string error;
+    tf.reset(Trimmer::create());
+    bool b = tf->setup(trimmer, error);
+    if (!b) {
+      std::cout << error << std::endl;
+      return -1;
+    }
+  }
+
+  for(; obj_it->valid(); obj_it->next()) {
+    f = &*actual_formatter;
+    if (!trimmer.empty()) {
+      f = tf->attach(f);
+    }
+    if (!selector.empty()) {
+      matched = false;
+      buffer = bf->attach(f);
+      f = ff->attach(buffer, [&](){matched = true;});
+    }
+    std::string key = obj_it->key();
+    ghobject_t oid;
+    bool is_oid = *key.rbegin() == 'o';
+    if (is_oid)
+      dump_object(key, obj_it->value());
+    else
+      dump_extent(key, obj_it->value());
```

Comment:
IMO it's better not to dump extent shards separately (they are just physical parts of a single logical entity - Onode) Instead onode should load everything it needs and dump itself as a whole logical entity.